### PR TITLE
S_newONCEOP: Rejig the op_next padop (Closes #18630)

### LINKED ISSUE #18630: Returning state hash-assign risks "Bizarre copy of HASH in subroutine exit"
**Description**
This code gives the error `Bizarre copy of HASH in subroutine exit`

**Steps to Reproduce**
```
sub tt {state %h=(a=>1)}
print tt; # prints a1
print tt; # error `Bizarre copy of HASH in subroutine exit`
```

**Expected behavior**
both lines to print "a1"

**Perl configuration**

```
Summary of my perl5 (revision 5 version 32 subversion 1) configuration:
   
  Platform:
    osname=darwin
    osvers=20.2.0
    archname=darwin-2level
    uname='darwin yarys-apple-sauce 20.2.0 darwin kernel version 20.2.0: wed dec 2 20:39:59 pst 2020; root:xnu-7195.60.75~1release_x86_64 x86_64 '
    config_args='-de -Dprefix=/Users/hluchan/perl5/perlbrew/perls/perl-5.32.1 -Aeval:scriptdir=/Users/hluchan/perl5/perlbrew/perls/perl-5.32.1/bin'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
    bincompat5005=undef
  Compiler:
    cc='cc'
    ccflags ='-fno-common -DPERL_DARWIN -mmacosx-version-min=11.1 -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -I/opt/local/include -DPERL_USE_SAFE_PUTENV'
    optimize='-O3'
    cppflags='-fno-common -DPERL_DARWIN -mmacosx-version-min=11.1 -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -I/opt/local/include'
    ccversion=''
    gccversion='Apple LLVM 12.0.0 (clang-1200.0.32.28)'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags =' -mmacosx-version-min=11.1 -fstack-protector-strong -L/usr/local/lib -L/opt/local/lib'
    libpth=/usr/local/lib /Library/Developer/CommandLineTools/usr/lib/clang/12.0.0/lib /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/lib /Library/Developer/CommandLineTools/usr/lib /opt/local/lib /usr/lib
    libs=-lgdbm
    perllibs=
    libc=
    so=dylib
    useshrplib=false
    libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=bundle
    d_dlsymun=undef
    ccdlflags=' '
    cccdlflags=' '
    lddlflags=' -mmacosx-version-min=11.1 -bundle -undefined dynamic_lookup -L/usr/local/lib -L/opt/local/lib -fstack-protector-strong'


Characteristics of this binary (from libperl): 
  Compile-time options:
    HAS_TIMES
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_USE_SAFE_PUTENV
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
  Built under darwin
  Compiled at Jan 25 2021 12:29:31
  %ENV:
    PERLBREW_HOME="/Users/hluchan/.perlbrew"
    PERLBREW_MANPATH="/Users/hluchan/perl5/perlbrew/perls/perl-5.32.1/man"
    PERLBREW_PATH="/Users/hluchan/perl5/perlbrew/bin:/Users/hluchan/perl5/perlbrew/perls/perl-5.32.1/bin"
    PERLBREW_PERL="perl-5.32.1"
    PERLBREW_ROOT="/Users/hluchan/perl5/perlbrew"
    PERLBREW_SHELLRC_VERSION="0.87"
    PERLBREW_VERSION="0.87"
    PERL_CPANM_OPT="--mirror https://hwte.apple.com/repo/CPAN --mirror-only"
  @INC:
    /Users/hluchan/perl5/perlbrew/perls/perl-5.32.1/lib/site_perl/5.32.1/darwin-2level
    /Users/hluchan/perl5/perlbrew/perls/perl-5.32.1/lib/site_perl/5.32.1
    /Users/hluchan/perl5/perlbrew/perls/perl-5.32.1/lib/5.32.1/darwin-2level
    /Users/hluchan/perl5/perlbrew/perls/perl-5.32.1/lib/5.32.1
```

### PR DESCRIPTION
**Pull request by richardleach**
Two related changes pertaining to the padop executed when an OP_ONCE is encountered for the nth time.

* _S_newONCEOP_: correctly copy array/hash at subroutine exit 
Should a _state_ array or hash be declared immediately prior to
returning, the error `Bizarre copy of HASH in subroutine exit` error
would be triggered. Now, the array or hash is returned without error.
Closes #18630

* _S_newONCEOP_/_Perl_scalarvoid_: skip over padops uselss in void context
Rejigs how the nth-time padop is handled in void context to avoid the
warning "Useless use of private variable in void context". The OPf_STATE
flag is removed and the op_next chain now explicitly skips the padop.

---------------------------------------------------------------------------------
* This set of changes requires a perldelta entry, and it is included.

### CODE REVIEW

File: ext/B/t/optree_varinit.t [hunk 1]
```diff
@@ -405,8 +405,6 @@ checkOptree ( name	=> 'void context state',
 # 3  <|> once(other->4)[$:2,3] vK/1
```
Comment (tonycoz) [inline ext/B/t/optree_varinit.t:405]: "uselss" in commit message

**Comment by tonycoz** [comment]
An array in scalar context is normally the number of elements in the array, but that doesn't appear to be the case for the trailing `state @foo`:
```
$ ./perl -Ilib -E 'sub f { state @x = @_; } say f(3,2,1); say scalar f()'
321
1
$ ./perl -Ilib -E 'sub f { state @x = @_; } say scalar f(4,2,1); say scalar f()'
3
1
```

**Comment by richardleach** [comment]
Thanks, I'll look into it.

I see that in:

```
./perl -Ilib -E 'sub f { state @x = @_; @x; } say f(3,2,1); say scalar f()'
321
3
$ perl -E 'sub f { my @tto =(3,2,1) } say f(3,2,1); say scalar f()'
321
3
$ perl -E 'sub f { my @tto =(3,2,1); @tto } say f(3,2,1); say scalar f()'
321
3
```
Though not in:
```
$ ./perl -Ilib -E 'sub f { (3,2,1) } say f(3,2,1); say scalar f()'
321
1
```

**Comment by richardleach** [comment]
Looks like it comes down to clobbering (or not) `OPf_WANT`. I'll update the PR shortly.

**Comment by richardleach** [comment]
Ok, I think it has the correct behaviour now.

**Review by tonycoz — APPROVED** [review]

**Pull request merged**